### ISSUE TITLE
Improve memory space efficiency and add cuda optimizations

### ISSUE THREAD

**Issue by bkuen** [jualat/CleanRLHF#48]
### The problem

- At the moment, not all tensors are bound correctly to device
- We calculate unnecessary gradients which lead to high (GPU) memory consumption

### The solution / the feature

- Add additional .to(device) calls
- Use with torch.no_grad() to avoid calculating unnecessary gradients
